--- ./source/scriptinterface/ScriptContext.cpp.orig	2021-08-16 21:29:42.364933739 +0100
+++ ./source/scriptinterface/ScriptContext.cpp	2021-08-16 21:30:56.222150308 +0100
@@ -143,7 +143,7 @@
 	// Schedule the zone for GC, which will destroy the realm.
 	if (JS::IsIncrementalGCInProgress(m_cx))
 		JS::FinishIncrementalGC(m_cx, JS::GCReason::API);
-	JS::PrepareZoneForGC(js::GetRealmZone(realm));
+	JS::PrepareZoneForGC(m_cx, js::GetRealmZone(realm));
 	m_Realms.remove(realm);
 }
 
@@ -261,5 +261,5 @@
 void ScriptContext::PrepareZonesForIncrementalGC() const
 {
 	for (JS::Realm* const& realm : m_Realms)
-		JS::PrepareZoneForGC(js::GetRealmZone(realm));
+		JS::PrepareZoneForGC(m_cx, js::GetRealmZone(realm));
 }
