--- ./neo/sys/sdl/sdl_glimp.cpp~	2020-11-12 09:17:05.000000000 +0000
+++ ./neo/sys/sdl/sdl_glimp.cpp	2020-11-12 09:24:01.667646782 +0000
@@ -113,6 +113,10 @@
 	Uint32 flags = SDL_WINDOW_OPENGL | SDL_WINDOW_RESIZABLE;
 	// DG end
 
+#if SDL_VERSION_ATLEAST(2, 0, 0)
+	flags |= SDL_WINDOW_ALLOW_HIGHDPI;
+#endif
+
 	if( parms.fullScreen )
 	{
 		flags |= SDL_WINDOW_FULLSCREEN;
@@ -298,7 +302,7 @@
 		}
 
 		// RB begin
-		SDL_GetWindowSize( window, &glConfig.nativeScreenWidth, &glConfig.nativeScreenHeight );
+		SDL_GL_GetDrawableSize( window, &glConfig.nativeScreenWidth, &glConfig.nativeScreenHeight );
 		// RB end
 
 		glConfig.isFullscreen = ( SDL_GetWindowFlags( window ) & SDL_WINDOW_FULLSCREEN ) == SDL_WINDOW_FULLSCREEN;
--- ./neo/sys/sdl/sdl_vkimp.cpp~	2020-11-12 09:17:05.000000000 +0000
+++ ./neo/sys/sdl/sdl_vkimp.cpp	2020-11-12 09:33:34.336460831 +0000
@@ -143,6 +143,10 @@
 	Uint32 flags = SDL_WINDOW_VULKAN | SDL_WINDOW_RESIZABLE;
 	// DG end
 
+#if SDL_VERSION_ATLEAST(2, 0, 8)
+	flags |= SDL_WINDOW_ALLOW_HIGHDPI;
+#endif
+
 	if( parms.fullScreen )
 	{
 		flags |= SDL_WINDOW_FULLSCREEN;
@@ -275,7 +279,11 @@
 		}
 		vkcontext.sdlWindow = window;
 		// RB begin
+#if SDL_VERSION_ATLEAST(2, 0, 8)
+		SDL_Vulkan_GetDrawableSize( window, &glConfig.nativeScreenWidth, &glConfig.nativeScreenHeight );
+#else
 		SDL_GetWindowSize( window, &glConfig.nativeScreenWidth, &glConfig.nativeScreenHeight );
+#endif
 		// RB end
 
 		glConfig.isFullscreen = ( SDL_GetWindowFlags( window ) & SDL_WINDOW_FULLSCREEN ) == SDL_WINDOW_FULLSCREEN;
--- ./neo/sys/sdl/sdl_glimp.cpp~	2020-11-12 09:58:36.000000000 +0000
+++ ./neo/sys/sdl/sdl_glimp.cpp	2020-11-12 10:40:49.155380277 +0000
@@ -556,8 +556,10 @@
 
 	glConfig.isFullscreen = parms.fullScreen;
 	glConfig.isStereoPixelFormat = parms.stereo;
+#if !SDL_VERSION_ATLEAST(2, 0, 0)
 	glConfig.nativeScreenWidth = parms.width;
 	glConfig.nativeScreenHeight = parms.height;
+#endif
 	glConfig.displayFrequency = parms.displayHz;
 	glConfig.multisamples = parms.multiSamples;
 
--- ./neo/sys/sdl/sdl_vkimp.cpp~	2020-11-12 10:42:01.000000000 +0000
+++ ./neo/sys/sdl/sdl_vkimp.cpp	2020-11-12 10:56:46.495687272 +0000
@@ -455,8 +455,10 @@
 
 	glConfig.isFullscreen = parms.fullScreen;
 	glConfig.isStereoPixelFormat = parms.stereo;
+#if !SDL_VERSION_ATLEAST(2, 0, 8)
 	glConfig.nativeScreenWidth = parms.width;
 	glConfig.nativeScreenHeight = parms.height;
+#endif
 	glConfig.displayFrequency = parms.displayHz;
 	glConfig.multisamples = parms.multiSamples;
 
